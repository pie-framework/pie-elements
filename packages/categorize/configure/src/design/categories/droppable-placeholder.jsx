import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import ChoicePreview from './choice-preview';
import { DropTarget } from 'react-dnd';
import { uid, PlaceHolder } from '@pie-lib/drag';
import debug from 'debug';

const log = debug('@pie-element:categorize:configure');

const Helper = withStyles(() => ({
  helper: {
    display: 'flex',
    alignItems: 'center',
    fontSize: '12px',
    color: 'rgba(0,0,0,0.4)',
    width: '100%',
    height: '100%',
  },
}))(({ classes }) => <div className={classes.helper}>Drag your correct answers here</div>);

const Previews = ({ alternateResponseIndex, category, choices, onDeleteChoice }) => (
  <React.Fragment>
    {choices.map(
      (c, index) =>
        c && (
          <ChoicePreview
            alternateResponseIndex={alternateResponseIndex}
            category={category}
            choice={c}
            key={index}
            choiceIndex={index}
            onDelete={(choice) => onDeleteChoice(choice, index)}
          />
        ),
    )}
  </React.Fragment>
);

Previews.propTypes = {
  alternateResponseIndex: PropTypes.number,
  category: PropTypes.object,
  choices: PropTypes.array,
  onDeleteChoice: PropTypes.func,
};

export class DroppablePlaceHolder extends React.Component {
  static propTypes = {
    alternateResponseIndex: PropTypes.number,
    category: PropTypes.object,
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    connectDropTarget: PropTypes.func.isRequired,
    choices: PropTypes.array,
    onDropChoice: PropTypes.func.isRequired,
    onMoveChoice: PropTypes.func,
    isOver: PropTypes.bool,
    onDeleteChoice: PropTypes.func,
    categoryId: PropTypes.string.isRequired,
  };

  static defaultProps = {};
  render() {
    const { alternateResponseIndex, isOver, category, choices, classes, className, connectDropTarget, onDeleteChoice } =
      this.props;

    return connectDropTarget(
      <div className={classNames(classes.droppablePlaceholder, className)}>
        <PlaceHolder isOver={isOver} className={classes.placeHolder}>
          {(choices || []).length === 0 ? (
            <Helper />
          ) : (
            <Previews
              alternateResponseIndex={alternateResponseIndex}
              category={category}
              choices={choices}
              onDeleteChoice={onDeleteChoice}
            />
          )}
        </PlaceHolder>
      </div>,
    );
  }
}
const styles = () => ({
  droppablePlaceholder: {
    minHeight: '100px',
  },
  placeHolder: {
    width: '100%',
    minHeight: '100px',
    height: 'auto',
  },
});

const Styled = withStyles(styles)(DroppablePlaceHolder);

export const spec = {
  drop: (props, monitor) => {
    log('[drop] props: ', props);
    const item = monitor.getItem();

    if (item.from && item.alternateResponseIndex === props.alternateResponseIndex) {
      props.onMoveChoice(item.choiceId, item.from, props.categoryId, item.choiceIndex, item.alternateResponseIndex);
    } else {
      props.onDropChoice(item, props.categoryId);
    }
  },
  canDrop: (props /*, monitor*/) => {
    return !props.disabled;
  },
};

const WithTarget = DropTarget(
  ({ uid }) => uid,
  spec,
  (connect, monitor) => ({
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver(),
  }),
)(Styled);

export default uid.withUid(WithTarget);
