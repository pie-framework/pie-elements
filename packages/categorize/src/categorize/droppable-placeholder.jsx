import React from 'react';
import { PlaceHolder } from '@pie-lib/pie-toolbox/drag';
import PropTypes from 'prop-types';
import { DropTarget } from 'react-dnd';
import { uid } from '@pie-lib/pie-toolbox/drag';
import debug from 'debug';

const log = debug('@pie-ui:categorize:droppable-placeholder');

export class DroppablePlaceholder extends React.Component {
  static propTypes = {
    choiceBoard: PropTypes.bool,
    connectDropTarget: PropTypes.func.isRequired,
    isOver: PropTypes.bool,
    children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
    className: PropTypes.string,
    grid: PropTypes.object,
    disabled: PropTypes.bool,
    minRowHeight: PropTypes.string,
  };
  render() {
    const {
      children,
      connectDropTarget,
      isOver,
      className,
      grid,
      disabled,
      choiceBoard,
      minRowHeight
    } = this.props;

    return connectDropTarget(
      <div style={{ flex: 1, minHeight: minRowHeight || '10px' }}>
        <PlaceHolder
          className={className}
          isOver={isOver}
          grid={grid}
          disabled={disabled}
          choiceBoard={choiceBoard}
          isCategorize
        >
          {children}
        </PlaceHolder>
      </div>,
    );
  }
}

export const spec = {
  drop: (props, monitor) => {
    log('[drop] props: ', props);
    const item = monitor.getItem();
    props.onDropChoice(item);
  },
  canDrop: (props /*, monitor*/) => {
    return !props.disabled;
  },
};

const WithTarget = DropTarget(
  ({ uid }) => uid,
  spec,
  (connect, monitor) => ({
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver(),
  }),
)(DroppablePlaceholder);

export default uid.withUid(WithTarget);
