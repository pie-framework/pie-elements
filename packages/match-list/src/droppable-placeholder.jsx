import React from 'react';
import { PlaceHolder, uid } from '@pie-lib/drag';
import PropTypes from 'prop-types';
import { DropTarget } from 'react-dnd';

export class DroppablePlaceholder extends React.Component {
  static propTypes = {
    connectDropTarget: PropTypes.func.isRequired,
    isOver: PropTypes.bool,
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node
    ]).isRequired,
    disabled: PropTypes.bool
  };
  render() {
    const {
      children,
      connectDropTarget,
      isOver,
      disabled
    } = this.props;

    return connectDropTarget(
      <div style={{ flex: 1 }}>
        <PlaceHolder
          disabled={disabled}
          isOver={isOver}
        >
          {children}
        </PlaceHolder>
      </div>
    );
  }
}

export const spec = {
    canDrop: (props /*, monitor*/) => {
       return !props.disabled;
      },
  drop: (props) => {
    log('[drop] props: ', props);
  }
};

const WithTarget = DropTarget('Answer', spec, (connect, monitor) => ({
  connectDropTarget: connect.dropTarget(),
  isOver: monitor.isOver(),
}))(DroppablePlaceholder);

export default WithTarget;
