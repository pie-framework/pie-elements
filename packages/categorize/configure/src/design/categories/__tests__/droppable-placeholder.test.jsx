import { shallow } from 'enzyme';
import React from 'react';

import { DroppablePlaceHolder, spec } from '../droppable-placeholder';

describe('DroppablePlaceholder', () => {
  let w;
  let connectDropTarget = jest.fn((o) => o);
  let onDropChoice = jest.fn();
  let onDeleteChoice = jest.fn();
  const wrapper = (extras) => {
    const defaults = {
      classes: {},
      className: 'className',
      onDeleteChoice,
      onDropChoice,
      connectDropTarget,
      categoryId: '1',
      choices: [{ id: '1', content: 'content' }],
    };
    const props = {
      ...defaults,
      ...extras,
    };
    return shallow(<DroppablePlaceHolder {...props} />);
  };
  describe('snapshot', () => {
    it('renders', () => {
      w = wrapper();
      expect(w).toMatchSnapshot();
    });

    it('renders helper', () => {
      w = wrapper({ choices: [] });
      expect(w).toMatchSnapshot();
    });
  });
});

describe('spec', () => {
  describe('drop', () => {
    it('calls onDropChoice', () => {
      const props = {
        onDropChoice: jest.fn(),
        categoryId: '1',
      };

      const item = { id: '2' };
      const monitor = {
        getItem: jest.fn().mockReturnValue(item),
      };
      spec.drop(props, monitor);
      expect(props.onDropChoice).toBeCalledWith(item, props.categoryId);
    });
  });
  describe('canDrop', () => {
    it('returns true if !disabled', () => {
      expect(spec.canDrop({ disabled: false })).toEqual(true);
    });
    it('returns false if disabled', () => {
      expect(spec.canDrop({ disabled: true })).toEqual(false);
    });
  });
});
