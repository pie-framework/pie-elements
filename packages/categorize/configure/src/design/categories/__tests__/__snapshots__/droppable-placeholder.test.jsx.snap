// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DroppablePlaceholder snapshot renders 1`] = `
<div
  className="className"
>
  <WithStyles(PlaceHolder)>
    <Previews
      choices={
        Array [
          Object {
            "content": "content",
            "id": "1",
          },
        ]
      }
      onDeleteChoice={[MockFunction]}
    />
  </WithStyles(PlaceHolder)>
</div>
`;

exports[`DroppablePlaceholder snapshot renders helper 1`] = `
<div
  className="className"
>
  <WithStyles(PlaceHolder)>
    <WithStyles(Component) />
  </WithStyles(PlaceHolder)>
</div>
`;
